pkg/csi/service/vanilla/controller.go

```diff
@@ -2274,6 +2267,291 @@ func (c *controller) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRe
+				},
+			}
+			for {
+				queryResult, err := c.manager.VolumeManager.QuerySnapshots(ctx, snapshotQueryFilter)
+				if err != nil {
+					log.Errorf("querySnapshots failed. Err=%+v", err)
+					return nil, "", err
+				}
+				snapQueryEntries = append(snapQueryEntries, queryResult.Entries...)
+				// If cursor offset < total snapshot records found, then there are more snapshots
+				// to be queried, so continue the loop to fetch remaining records
+				if queryResult.Cursor.Offset >= queryResult.Cursor.TotalRecords {
+					break
+				}
+				// Assign cursor for next iteration based on previous query result
+				snapshotQueryFilter.Cursor = &queryResult.Cursor
+			}
+			//populate list of volume-ids to retrieve the volume size.
+			var volumeIds []cnstypes.CnsVolumeId
+			for _, queryResult := range snapQueryEntries {
+				if queryResult.Error != nil {
+					return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+						"faults are not expected when invoking QuerySnapshots without volume-id and snapshot-id, fault: %+v",
+						queryResult.Error.Fault)
+				}
+				volumeIds = append(volumeIds, queryResult.Snapshot.VolumeId)
+			}
+			// TODO: Retrieve Snapshot size directly from CnsQuerySnapshot once supported.
+			volumeDetails, err := utils.QueryVolumeDetailsUtil(ctx, c.manager.VolumeManager, volumeIds)
+			if err != nil {
+				log.Errorf("failed to retrieve volume details for volume-ids: %v, err: %+v", volumeIds, err)
+				return nil, "", err
+			}
+			CNSVolumeDetailsMap = append(CNSVolumeDetailsMap, volumeDetails)
+			CNSSnapshotsForListSnapshots = snapQueryEntries
+		}
+	}
+	// Process snapshot query result obtained from configured vCenter(s) and return final response
+	finalSnapEntries, nextToken, err = c.processQueryResultsListSnapshots(ctx, startingToken,
+		maxEntries, CNSSnapshotsForListSnapshots)
+	if err != nil {
+		return nil, csifault.CSIInternalFault, fmt.Errorf("error while processing query results for list "+
+			" snapshots, err: %v", err)
+	}
+	for _, queryResult := range finalSnapEntries {
+		snapshotCreateTimeInProto := timestamppb.New(queryResult.Snapshot.CreateTime)
+		csiSnapshotId := queryResult.Snapshot.VolumeId.Id +
+			common.VSphereCSISnapshotIdDelimiter + queryResult.Snapshot.SnapshotId.Id
+		// For every snapshot, find corresponding source volume details and fill the CSISnapshot response entry
+		for _, cnsVolumeDetailsMap := range CNSVolumeDetailsMap {
+			if _, ok := cnsVolumeDetailsMap[queryResult.Snapshot.VolumeId.Id]; !ok {
+				continue
+			}
+			csiSnapshotSize := cnsVolumeDetailsMap[queryResult.Snapshot.VolumeId.Id].SizeInMB * common.MbInBytes
+			csiSnapshotInfo := &csi.Snapshot{
+				SnapshotId:     csiSnapshotId,
+				SourceVolumeId: queryResult.Snapshot.VolumeId.Id,
+				CreationTime:   snapshotCreateTimeInProto,
+				SizeBytes:      csiSnapshotSize,
+				ReadyToUse:     true,
+			}
+			csiSnapshots = append(csiSnapshots, csiSnapshotInfo)
+			break
+		}
+	}
+	return csiSnapshots, nextToken, nil
```

Comment:
Can you drop a log line that states the number of results you have returned, also log the number of results you are yet to return

Reply:
done